src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -97,7 +97,14 @@ protected virtual void Dispose(bool disposing)
         {
             if (disposing)
             {
-                this.semaphore.Dispose();
+                // !!DO NOT!! dispose the semaphore, for the following reasons:
+                // 1. SemaphoreSlim.Dispose is not safe for concurrent use while something is waiting. Calling Dispose
+                //    requires tracking of all outstanding requests for enter/release.
+                // 2. SemaphoreSlim only allocates resources that could need to be be disposed if the
+                //    SemaphoreSlim.AvailableWaitHandle property is accessed. The semaphore field here is private and we
+                //    do not access this property.
```

Comment:
If `StreamJsonRpc` is throwing (or propagating) `ArgumentNullException` in response to a stream failure, I agree that behavior is undesirable. I hadn't noticed that that was the exception type. Can someone file an issue with on the vs-streamjsonrpc repo to track that?

Given `SemaphoreSlim` is not thread-safe for a `Dispose` call, then if nothing else we should document that `AsyncSemaphore` is not either. It was not originally intended to be more thread-safe than its underlying wrapped primitive.